File: 2019/01.rb
#!/usr/bin/env ruby
mass_values = File.readlines("#{__dir__}/01.txt").map(&:to_i)

part1 = 0
mass_values.each { |mass| part1 += mass / 3 - 2 }
puts "Part 1: #{part1}"

part2 = 0
mass_values.each do |mass|
  fuel = mass
  # _Could_ cache fuel values here to potentially save some computation, but...meh
  while fuel > 0
    fuel = fuel / 3 - 2
    part2 += fuel if fuel > 0
  end
end
puts "Part 2: #{part2}"

File: 2019/01.go
package main

import (
	"fmt"
	"main/helper"
	"strconv"
)

var cache map[int]int

func part2cost(mass int) int {
	if cost, ok := cache[mass]; ok {
		return cost
	}
	var fuel int = mass/3 - 2
	if fuel <= 0 {
		cache[mass] = 0
	} else {
		cache[mass] = fuel + part2cost(fuel)
	}
	return cache[mass]
}

func main() {
	cache = make(map[int]int)
	var part1, part2 int = 0, 0

	for line := range helper.GetInputLineChannel() {
		mass, _ := strconv.Atoi(line)
		part1 += mass/3 - 2
		part2 += part2cost(mass)
	}

	fmt.Printf("Part 1: %d\n", part1)
	fmt.Printf("Part 2: %d\n", part2)
}
